### PR TITLE
[mono][aot] Deferred failures code cleanup

### PR DESCRIPTION
**Pull request by kotlarmilos**
This PR adds `mono_` prefix for functions and types, and moves code related to class manipulation to the `mono/metadata` directory.

### CODE REVIEW

File: src/mono/mono/metadata/exception.c [hunk 1]
```diff
@@ -1458,3 +1458,63 @@ mono_error_convert_to_exception_handle (MonoError *error)
...
+ * Sets a deferred failure in the class and prints a warning message. 
+ * The deferred failure allows the runtime to attempt setting up the class layout at runtime.
+ *
+ * LOCKING: Acquires the loader lock.
+ *
+ * \returns FALSE
+ */
+gboolean
+mono_class_set_deferred_type_load_failure (MonoClass *klass, const char * fmt, ...)
+{
```
Comment (vargaz) [inline src/mono/mono/metadata/exception.c:1476]: These should be in class.c.

**Review by lambdageek — DISMISSED** [review]
Thanks!

File: src/mono/mono/utils/mono-error-internals.h [hunk 2]
```diff
@@ -320,12 +320,12 @@ void
 mono_error_set_first_argument (MonoError *oerror, const char *first_argument);
 
 typedef enum {
-    DEFERRED_FAILURE, // Used during AOT compilation to defer failure for execution
-    IMMEDIATE_FAILURE // Used during runtime to indicate that the failure should be reported
+    MONO_CLASS_LOADER_IMMEDIATE_FAILURE, // Used during runtime to indicate that the failure should be reported
+    MONO_CLASS_LOADER_DEFERRED_FAILURE // Used during AOT compilation to defer failure for execution
 } FailureType;
```
Comment (lambdageek) [inline src/mono/mono/utils/mono-error-internals.h:325]: nit: Mono prefix here too

File: src/mono/mono/utils/mono-error-internals.h [hunk 3]
```diff
@@ -320,12 +320,12 @@ void
...
 
 void
-set_failure_type (FailureType failure_type);
+mono_set_failure_type (FailureType failure_type);
```
Comment (lambdageek) [inline src/mono/mono/utils/mono-error-internals.h:328]: This declaration and `MonoFailureType` should be in `metadata-internals.h` or something like that...

Also maybe `mono_class_loader_set_failure_type` ?

**Review by lambdageek — APPROVED** [review]
a few more cleanup suggestions

File: src/mono/mono/metadata/metadata-internals.h [hunk 4]
```diff
@@ -682,6 +682,27 @@ typedef struct {
...
+ * @param fmt: printf-style error message string.
+ *
+ * The callback is responsible for processing the failure information provided by the @klass parameter and the error message format string @fmt.
+ * If a deferred failure occurs, the callback should return FALSE to let the AOT compiler proceed with the class layout setup.
+ * Otherwise, if the callback returns TRUE, it indicates that the failure should be reported.
+ *
+ * @returns: TRUE if the failure is handled and the runtime should not proceed with class setup, FALSE if the failure should be deferred for runtime class setup.
+ * 
+ */
+typedef gboolean (*TypeLoadFailureCallback)(MonoClass *klass, const char * fmt, ...) MONO_ATTR_FORMAT_PRINTF(2,3);
```
Comment (vargaz) [inline src/mono/mono/metadata/metadata-internals.h:702]: There is a MonoRuntimeCallbacks structure in object-internals.h which could include this.

**Comment by kotlarmilos** [comment]
I've added 

```
TypeLoadFailureCallback mono_class_set_deferred_type_load_failure_callback = mono_class_set_type_load_failure;
```
 to `class.c`. Let me know if you have a better idea on where to initialize it.

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/metadata/metadata-internals.h (hunk 4 above)
Comment (kotlarmilos) [reply src/mono/mono/metadata/metadata-internals.h]: I like the idea. I've moved it and added checks if the callback is initialized, like
`if (mono_get_runtime_callbacks ()->mono_class_set_deferred_type_load_failure_callback)`

**Comment by kotlarmilos** [comment]
The failures are known and shouldn't be related.

**Pull request merged**